cmd/terraformer/app/terraformer.go

```diff
@@ -0,0 +1,138 @@
+		Long: `terraformer executes terraform commands inside a Kubernetes cluster and handles Pod lifecycle event (e.g. shutdown signals).
+It reads and stores terraform config and state from/to Kubernetes resources (ConfigMaps and Secrets).
+Also, it continuously updates the state ConfigMap by running a file watcher and updating the ConfigMap as soon as the state file changes.`,
+		Version: version.Version,
+		Hidden:  true,
+
+		PersistentPreRun: func(cmd *cobra.Command, args []string) {
+			runtimelog.SetLogger(logzap.New(
+				// use configuration passed via flags
+				logzap.UseFlagOptions(zapOpts),
+				// and overwrite some stuff
+				func(o *logzap.Options) {
+					if !o.Development {
+						encCfg := zap.NewProductionEncoderConfig()
+						// overwrite time encoding to human readable format for production logs
+						encCfg.EncodeTime = zapcore.ISO8601TimeEncoder
+						o.Encoder = zapcore.NewJSONEncoder(encCfg)
+					}
+
+					// don't print stacktrace for warning level logs
+					o.StacktraceLevel = zapcore.ErrorLevel
+				},
+			))
+
+			runtimelog.Log.Info("Starting terraformer...", "version", version.Version)
+			cmd.Flags().VisitAll(func(flag *pflag.Flag) {
+				runtimelog.Log.Info(fmt.Sprintf("FLAG: --%s=%s", flag.Name, flag.Value))
+			})
+		},
+
+		RunE: func(cmd *cobra.Command, args []string) error {
+			// root command called without any terraform command is invalid and should exit with error code
+			return errors.New("no command was specified")
+		},
+	}
+	cmd.SetUsageTemplate(cosmeticUsageTemplate)
+
+	// setup a subcommand for every supported terraform command
+	for command := range terraformer.SupportedCommands {
+		addSubcommand(cmd, command, tfOpts)
+	}
+
+	// setup flags
+	tfOpts.AddFlags(cmd.PersistentFlags())
+	zapOpts.BindFlags(flag.CommandLine)
+	cmd.PersistentFlags().AddGoFlagSet(flag.CommandLine)
+
+	return cmd
+}
+
+func addSubcommand(cmd *cobra.Command, command terraformer.Command, opts *terraformercmd.Options) {
+	cmd.AddCommand(&cobra.Command{
+		Use:   string(command),
+		Short: fmt.Sprintf("execute `terraform %s`", command),
+		Long:  fmt.Sprintf("terraformer %s executes the terraform %s command with the given configuration", command, command),
+		Args:  cobra.NoArgs,
+
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if err := opts.Complete(); err != nil {
+				return err
+			}
+
+			// don't output usage on further errors raised during terraform execution
+			cmd.SilenceUsage = true
+			// further errors will be logged properly, don't duplicate
+			cmd.SilenceErrors = true
+
+			tf, err := terraformer.NewDefaultTerraformer(opts.Completed())
+			if err != nil {
+				return err
+			}
+
+			return tf.Run(command)
+		},
+	})
+}
+
+// custom usage template:
+// - exclude `terraformer [flags]` if command is hidden (e.g. root command)
+// - indicate [flags] after `terraformer [command]` (for subcommands)
+const cosmeticUsageTemplate = `Usage:{{if and .Runnable (not .Hidden)}}
```

Comment:
Can you explain when and how this is used?

Reply:
It is used instead of the default cobra usage template, i.e. if `terraformer help` or `terraformer [command] -h` is executed.

It differs only slightly from the default template. E.g. for `terraformer help` it puts out 
```
...
Usage:
  terraformer [command] [flags]
...
```
instead of 
```
...
Usage:
  terraformer [flags]
  terraformer [command]
...
```

So as the name suggests, this is only a minor cosmetic enhancement 😄

Reply:
Okay, thanks for explaining. I especially missed the `Examples:` section being printed in the help but that's probably because you don't have any examples configured. Never mind, I mainly asked out of curiosity reasons :)

Reply:
Specifying an example is probably a good idea anyways, so I added one for each command :)